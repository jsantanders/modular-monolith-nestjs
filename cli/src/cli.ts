#!/usr/bin/env node

import * as yargs from 'yargs';
import { AppConfig } from '../../backend/src/config/app.config';
import { getLogger } from '@nestjs-bff/backend/lib/shared/logging/logging.shared.module';

// global setup
AppConfig.appName = 'nestjs-bff-cli';

global.nestjs_bff = { config: AppConfig };
const bffLogger = getLogger();

// prettier-ignore
// tslint:disable-next-line:no-unused-expression
yargs
  .env('BFFCLI')
  .commandDir('commands', {
    extensions: AppConfig.nodeEnv === 'dev' ? ['js', 'ts'] : ['js'],
    visit: (commandObject) => {
      commandObject.loggerService = bffLogger;
      return commandObject;
    },
  })
  .demandCommand(1)
  .help()
  .version()
  .argv;
