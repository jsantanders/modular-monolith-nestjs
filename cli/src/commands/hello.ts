import { prompt as ask } from 'inquirer';
import * as yargs from 'yargs'; // eslint-disable-line no-unused-vars
import { LoggerSharedService } from '@nestjs-bff/backend/lib/shared/logging/logger.shared.service';

export let loggerService: LoggerSharedService = null; // This needs to be initialized by the commandLoader

async function askName(): Promise<string> {
  loggerService.log(':wave:  Hello stranger!');
  const { name } = await ask<IParams>([
    {
      type: 'input',
      name: 'name',
      message: 'What is your name?',
    },
  ]);
  return name;
}

export interface IParams {
  name?: string;
}

// Supply exports to satisfy Yargs command pattern: command, desc, builder, and handler
export const command = 'hello';
export const desc = `Let's get to know each other`;
export const builder: { [key: string]: yargs.Options } = {
  name: { type: 'string', required: false, description: 'your name' },
};
export async function handler({ name }: IParams) {
  loggerService.log(`Oh, nice to meet you, ${name || (await askName())}!`);
}
