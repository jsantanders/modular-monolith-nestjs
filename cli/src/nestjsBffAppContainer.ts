import { INestApplication } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { CliAppModule } from '../../backend/src/app/host/cli/cliapp.module';

export class NestjsBffAppContainer {
  public static appInstance: INestApplication;

  public static async ensureInitialized(): Promise<void> {
    if (!NestjsBffAppContainer.appInstance)
      await NestFactory.create(CliAppModule).then((nestApp: INestApplication) => {
        this.appInstance = nestApp;
      });
  }
}
