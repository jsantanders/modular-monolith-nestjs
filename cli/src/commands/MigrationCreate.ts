import * as yargs from 'yargs'; // eslint-disable-line no-unused-vars
import { MigrationsSharedService } from '@nestjs-bff/backend/lib/shared/migrations/migrations.shared.service';
import { NestjsBffAppContainer } from '../nestjsBffAppContainer';
import { LoggerSharedService } from '@nestjs-bff/backend/lib/shared/logging/logger.shared.service';

export let loggerService: LoggerSharedService; // This needs to be initialized by the commandLoader
export interface IParams {
  migrationName: string;
}

// Supply exports to satisfy Yargs command pattern: command, desc, builder, and handler
export const command = 'migration-create <migration-name>';
export const desc = `Creates a new migration file`;
export const builder: { [key: string]: yargs.Options } = {
  migrationName: { type: 'string', required: true, description: 'migration name' },
};
export async function handler({ migrationName }: IParams) {
  await NestjsBffAppContainer.ensureInitialized().then(() => {
    const migrationService = NestjsBffAppContainer.appInstance.get(MigrationsSharedService);
    migrationService.create('', migrationName).then(() => {
      process.exit(0);
    });
  });
}
