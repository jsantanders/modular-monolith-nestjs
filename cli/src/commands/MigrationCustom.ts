import * as yargs from 'yargs'; // eslint-disable-line no-unused-vars
import { MigrationsSharedService } from '@nestjs-bff/backend/lib/shared/migrations/migrations.shared.service';
import { NestjsBffAppContainer } from '../nestjsBffAppContainer';
import { LoggerSharedService } from '@nestjs-bff/backend/lib/shared/logging/logger.shared.service';

export let loggerService: LoggerSharedService; // This needs to be initialized by the commandLoader
export interface IParams {
  filename: string;
  direction: string;
}

// Supply exports to satisfy Yargs command pattern: command, desc, builder, and handler
export const command = 'migration-custom';
export const desc = `Runs a custom migration`;
export const builder: { [key: string]: yargs.Options } = {
  filename: { type: 'string', required: true, description: 'custom migration file' },
  direction: { type: 'string', required: true, description: 'the migration direction to run: up or down' },
};
export async function handler({ filename, direction }: IParams) {
  await NestjsBffAppContainer.ensureInitialized().then(() => {
    const migrationService = NestjsBffAppContainer.appInstance.get(MigrationsSharedService);
    migrationService
      .runCustomMigration(direction, filename)
      .then(() => {
        process.exit(0);
      })
      .catch(() => {
        process.exit(0);
      });
  });
}
