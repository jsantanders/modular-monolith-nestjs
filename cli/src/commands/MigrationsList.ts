import * as yargs from 'yargs'; // eslint-disable-line no-unused-vars
import { MigrationsSharedService } from '@nestjs-bff/backend/lib/shared/migrations/migrations.shared.service';
import { NestjsBffAppContainer } from '../nestjsBffAppContainer';
import { LoggerSharedService } from '@nestjs-bff/backend/lib/shared/logging/logger.shared.service';

export let loggerService: LoggerSharedService; // This needs to be initialized by the commandLoader
export interface IParams {}

// Supply exports to satisfy Yargs command pattern: command, desc, builder, and handler
export const command = 'migrations-list';
export const desc = `Lists all migrations and their current state`;
export const builder: { [key: string]: yargs.Options } = {};
export async function handler(argv: any) {
  await NestjsBffAppContainer.ensureInitialized().then(() => {
    const migrationService = NestjsBffAppContainer.appInstance.get(MigrationsSharedService);
    migrationService.list('').then(() => {
      process.exit(0);
    });
  });
}
