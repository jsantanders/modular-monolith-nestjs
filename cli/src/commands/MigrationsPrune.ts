import * as yargs from 'yargs'; // eslint-disable-line no-unused-vars
import { MigrationsSharedService } from '@nestjs-bff/backend/lib/shared/migrations/migrations.shared.service';
import { NestjsBffAppContainer } from '../nestjsBffAppContainer';
import { LoggerSharedService } from '@nestjs-bff/backend/lib/shared/logging/logger.shared.service';

export let loggerService: LoggerSharedService; // This needs to be initialized by the commandLoader
export interface IParams {}

// Supply exports to satisfy Yargs command pattern: command, desc, builder, and handler
export const command = 'migrations-prune';
export const desc =
  'Allows you to delete extraneous migrations by removing extraneous local migration files/database migrations.';
export const builder: { [key: string]: yargs.Options } = {};
export async function handler(argv: any) {
  await NestjsBffAppContainer.ensureInitialized().then(() => {
    const migrationService = NestjsBffAppContainer.appInstance.get(MigrationsSharedService);
    migrationService.prune('', true).then(() => {
      process.exit(0);
    });
  });
}
