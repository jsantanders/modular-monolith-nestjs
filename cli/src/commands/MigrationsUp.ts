import * as yargs from 'yargs'; // eslint-disable-line no-unused-vars
import { MigrationsSharedService } from '@nestjs-bff/backend/lib/shared/migrations/migrations.shared.service';
import { NestjsBffAppContainer } from '../nestjsBffAppContainer';
import { LoggerSharedService } from '@nestjs-bff/backend/lib/shared/logging/logger.shared.service';

export let loggerService: LoggerSharedService; // This needs to be initialized by the commandLoaderr
export interface IParams {
  migrationName: string;
}

// Supply exports to satisfy Yargs command pattern: command, desc, builder, and handler
export const command = 'migration-up';
export const desc = `Runs the UP script for all migrations up to and including the specified migration`;
export const builder: { [key: string]: yargs.Options } = {
  migrationName: { type: 'string', required: true, description: 'migration name' },
};
export async function handler({ migrationName }: IParams) {
  await NestjsBffAppContainer.ensureInitialized().then(() => {
    const migrationService = NestjsBffAppContainer.appInstance.get(MigrationsSharedService);
    migrationService
      .runMigration('up', migrationName)
      .then(() => {
        process.exit(0);
      })
      .catch(() => {
        process.exit(0);
      });
  });
}
