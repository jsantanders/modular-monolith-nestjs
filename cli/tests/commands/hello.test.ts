import { handler as hello } from '../../src/commands/hello';
import { runWithAnswers, ENTER } from '../utils';
import * as utils from '../../src/utils';

// silence outputs
jest.mock('inquirer/lib/utils/screen-manager');
jest.mock('../../src/utils');

const logInfo = jest.spyOn(utils, 'logInfo');
const renderQuestion = jest.spyOn(
  require('inquirer/lib/utils/screen-manager').prototype,
  'render'
);

afterEach(() => {
  logInfo.mockReset();
  renderQuestion.mockReset();
});

describe('hello', () => {
  it('says hello with the right name, after asking it if not provided', async () => {
    await runWithAnswers(() => hello({}), ['gabro', ENTER]);
    expect(renderQuestion.mock.calls[0][0]).toMatchSnapshot();
    expect(logInfo).toHaveBeenCalledTimes(2);
    expect(logInfo.mock.calls[0][0]).toMatchSnapshot();
    expect(logInfo.mock.calls[1][0]).toMatchSnapshot();
  });

  it('says hello with the right name, without asking it if provided', async () => {
    await runWithAnswers(() => hello({ name: 'gabro' }));
    expect(renderQuestion).not.toHaveBeenCalled();
    expect(logInfo).toHaveBeenCalledTimes(1);
    expect(logInfo.mock.calls[0][0]).toMatchSnapshot();
  });
});
