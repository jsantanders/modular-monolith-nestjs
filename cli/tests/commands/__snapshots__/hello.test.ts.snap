// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello says hello with the right name, after asking it if not provided 1`] = `"? What's your name? "`;

exports[`hello says hello with the right name, after asking it if not provided 2`] = `":wave:  Hello stranger!"`;

exports[`hello says hello with the right name, after asking it if not provided 3`] = `"Oh, nice to meet you, gabro!"`;

exports[`hello says hello with the right name, without asking it if provided 1`] = `"Oh, nice to meet you, gabro!"`;
